search.py

```diff
@@ -190,82 +234,179 @@ def merge_positions(positions1, positions2):
+                # curr1 = posting1.pointer.index
+            # elif posting2.pointer != None and posting2.pointer.doc_id < posting1.doc_id:
+                # curr2 = posting2.pointer.index
+            # # If we cannot jump, then we are left with the only option of incrementing the indexes one by one
+            # else:
+            if posting1.doc_id < posting2.doc_id:
+                curr1 += 1
             else:
-                if posting1.doc_id < posting2.doc_id:
-                    curr1 += 1
-                else:
-                    curr2 += 1
+                curr2 += 1
     return merged_list
 
 def parse_query(query):
-    phrasal_regex_pattern = '.*\"(.*)\".*'
-    if AND_KEYWORD in query or re.search(phrasal_regex_pattern, query):
-        return parse_boolean_query
-
+    terms_array, is_boolean_query = split_query(query)
+    if is_boolean_query:
+        return parse_boolean_query(terms_array)
     else:
-        return parse_free_text_query(query)
+        return parse_free_text_query(terms_array)
 
-def parse_boolean_query(query):
-    parse_by_arr = query.split(AND_KEYWORD)
-    stemmer = nltk.stem.porter.Stemmer()
-    parse_by_arr = [stemmer.stem(word.lower()) for word in parse_by_arr]
-    if not parse_by_arr:
-        return []
+def get_ranking_for_boolean_query(posting_list):
+    """
+    The scoring for boolean queries is going to follow CSS Specificity style
+    Title matches will be worth 20, court 10 and content 1 (numbers to be confirmed)
+    The overall relevance of the documents would be the sum of all these scores
+    Example: If the resultant posting list has two postings for doc_id xxx, with fields COURT and CONTENT
+    Then the resultant score is 11
+    """
+    title_score = 20
+    court_score = 10
+    content_score = 1
+
+    def get_boolean_query_scores(field):
+        if field == Field.TITLE:
+            return title_score
+        elif field == Field.COURT:
+            return court_score
+        else:
+            return content_score
+
+    scores = {}
+    for posting in posting_list.postings:
+        score = get_boolean_query_scores(posting.field)
+        if posting.doc_id not in scores:
+            scores[posting.doc_id] = score
+        else:
+            scores[posting.doc_id] += score
+
+    # Now we do the sorting
+    sorted_results = sorted(scores.items(), key=functools.cmp_to_key(comparator))
+
+    return [(doc_id, score) for doc_id, score in sorted_results]
 
-    first_term = parse_by_arr[0]
+def parse_boolean_query(terms):
+    """
+    Takes in the array of terms from the query
+    Returns the posting list of all the phrase
+    """
+    # First filter out all the AND keywords from the term array
+    filtered_terms = [term for term in terms if term != AND_KEYWORD]
+
+    filtered_terms = stem_query(process(terms))
+    # Get the posting list of the first word
+    first_term = filtered_terms[0]
     res_posting_list = None
     if " " in first_term:
```

Comment:
(if " " in first_term) would search strictly for " " in its specified order, can try doing ((first_term[0] == ") and (first_term[end] == ") ) or sth similar

Reply:
Oh we already removed the quotes in `split_query`